File: bindings/java-jni/src/main/java/org/maplibre/nativejni/resource/ResourceLoadingMethod.java
package org.maplibre.nativejni.resource;

/** Resource loading method copied from a native resource request. */
public final class ResourceLoadingMethod {
  public static final ResourceLoadingMethod ALL = new ResourceLoadingMethod(0, "ALL");
  public static final ResourceLoadingMethod CACHE_ONLY = new ResourceLoadingMethod(1, "CACHE_ONLY");
  public static final ResourceLoadingMethod NETWORK_ONLY =
      new ResourceLoadingMethod(2, "NETWORK_ONLY");

  private final int nativeValue;
  private final String name;

  private ResourceLoadingMethod(int nativeValue, String name) {
    this.nativeValue = nativeValue;
    this.name = name;
  }

  public int nativeValue() {
    return nativeValue;
  }

  public int rawValue() {
    return nativeValue;
  }

  public static ResourceLoadingMethod fromNative(int nativeValue) {
    return switch (nativeValue) {
      case 0 -> ALL;
      case 1 -> CACHE_ONLY;
      case 2 -> NETWORK_ONLY;
      default -> new ResourceLoadingMethod(nativeValue, null);
    };
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof ResourceLoadingMethod value && nativeValue == value.nativeValue;
  }

  @Override
  public int hashCode() {
    return Integer.hashCode(nativeValue);
  }

  @Override
  public String toString() {
    return name != null ? name : "ResourceLoadingMethod(" + nativeValue + ")";
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/resource/ResourceLoadingMethod.kt
package org.maplibre.nativeffi.resource

import kotlin.jvm.JvmInline

/**
 * Resource loading method copied from a native resource request.
 *
 * This is an open domain: a value may have no named constant here, so a `when` over this type needs
 * an `else` branch. Unknown values keep their raw [nativeValue].
 */
@JvmInline
public value class ResourceLoadingMethod(public val nativeValue: Int) {
  public companion object {
    public val ALL: ResourceLoadingMethod = ResourceLoadingMethod(0)
    public val CACHE_ONLY: ResourceLoadingMethod = ResourceLoadingMethod(1)
    public val NETWORK_ONLY: ResourceLoadingMethod = ResourceLoadingMethod(2)

    internal fun fromNative(nativeValue: UInt): ResourceLoadingMethod =
      fromNative(nativeValue.toInt())

    internal fun fromNative(nativeValue: Int): ResourceLoadingMethod =
      ResourceLoadingMethod(nativeValue)
  }
}
